### PR TITLE
Fix condition for SharedBuffer completion

### PR DESCRIPTION
**Pull request by nileema**
SharedBuffer should be marked complete when both conditions are satisfied:
- client has acknowledged all the pages
- client knows that there will be no more pages

Once the client sees all the pages and the no more pages signal,
explicitly send a delete to the SharedBuffer to indicate completion.

### CODE REVIEW

**Comment by martint** [comment]
Can you add an explanation of the observed behavior that can occur without this fix? (what we discovered when troubleshooting the issue caused by this)

File: presto-main/src/main/java/com/facebook/presto/execution/SharedBuffer.java [hunk 1]
```diff
@@ -467,11 +462,17 @@ public BufferResult getPages(long startingSequenceId, DataSize maxSize)
                 sequenceId = startingSequenceId;
             }
 
-            if (checkCompletion()) {
+            if (finished.get()) {
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/execution/SharedBuffer.java]: Use `isFinished`?

File: presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java [hunk 2]
```diff
@@ -556,8 +616,13 @@ public String toString()
                     .add("token", token)
                     .add("nextToken", nextToken)
                     .add("pagesSize", pages.size())
-                    .add("clientClosed", clientClosed)
+                    .add("clientComplete", clientComplete)
                     .toString();
         }
+
+        public boolean isClientComplete()
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java]: Place this with other getters like `getToken`

File: presto-main/src/test/java/com/facebook/presto/execution/TestSharedBuffer.java [hunk 3]
```diff
@@ -222,6 +224,7 @@ public void testSimple()
                 createPage(13)));
         assertQueueState(sharedBuffer, FIRST, DEFAULT_PARTITION, 8, 6, 8, 14, 0);
         assertBufferResultEquals(TYPES, getBufferResult(sharedBuffer, FIRST, 14, sizeOfPages(10), NO_WAIT), emptyResults(14, true));
+        sharedBuffer.abort(FIRST);
```
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/execution/TestSharedBuffer.java]: Add a comment above this line... something like `finish first queue`

File: presto-spi/src/main/java/com/facebook/presto/spi/StandardErrorCode.java [hunk 4]
```diff
@@ -45,6 +45,7 @@
     COMPILER_ERROR(0x0001_0007),
     WORKER_RESTARTED(0x0001_0008),
     SERVER_SHUTTING_DOWN(0x0001_0009),
+    CLOSE_TIMEOUT(0x0001_000A),
```
Comment (dain) [inline presto-spi/src/main/java/com/facebook/presto/spi/StandardErrorCode.java]: use `TOO_MANY_REQUESTS_FAILED` for this error

File: presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java [hunk 5]
```diff
@@ -257,19 +278,16 @@ private synchronized void initiateRequest()
                         .setUri(uri).build(),
                 new PageResponseHandler(blockEncodingSerde));
 
-        Futures.addCallback(future, new FutureCallback<PagesResponse>()
+        HttpResponseFuture<PagesResponse> resultFuture = (HttpResponseFuture<PagesResponse>) future;
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java]: I would do this the other way around.  Assign `executeAsync` result directly to the variable `resultFuture` and then in the second statement `this.future = resultFuture`. That way you don't need the cast.

File: presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java [hunk 6]
```diff
@@ -287,57 +305,94 @@ public void onSuccess(PagesResponse result)
...
             @Override
             public void onFailure(Throwable t)
             {
-                log.debug("Request to %s failed %s", uri, t);
+                checkNotHoldsLock();
 
-                if (Thread.holdsLock(HttpPageBufferClient.this)) {
-                    log.error("Can not handle callback while holding a lock on this");
-                }
+                log.debug("Request to %s failed %s", uri, t);
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java]: log before the check, incase there is a bug

File: presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java [hunk 7]
```diff
@@ -287,57 +305,94 @@ public void onSuccess(PagesResponse result)
...
+                Duration errorDuration = elapsedErrorDuration();
 
                 t = rewriteException(t);
-                if (t instanceof PrestoException) {
-                    clientCallback.clientFailed(HttpPageBufferClient.this, t);
-                }
-
-                Duration errorDuration = elapsedErrorDuration();
-                if (errorDuration.compareTo(minErrorDuration) > 0) {
+                if (!(t instanceof PrestoException) && errorDuration.compareTo(minErrorDuration) > 0) {
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java]: is this a change? Can't tell in the diff

File: presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java [hunk 8]
```diff
@@ -287,57 +305,94 @@ public void onSuccess(PagesResponse result)
...
+    }
 
-                increaseErrorDelay();
-
-                requestsFailed.incrementAndGet();
-                requestsCompleted.incrementAndGet();
+    private void sendDelete()
+    {
+        future = httpClient.executeAsync(prepareDelete().setUri(location).build(), createStatusResponseHandler());
+        HttpResponseFuture<StatusResponse> resultFuture = (HttpResponseFuture<StatusResponse>) future;
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java]: same here

**Comment by dain** [comment]
looks good

**Pull request merged**